File: src/main/java/com/babjo/whatdaybot/crawler/PeriodicRisingKeywordCrawler.java
package com.babjo.whatdaybot.crawler;

import static com.google.common.collect.ImmutableList.toImmutableList;

import java.io.IOException;
import java.time.Clock;
import java.time.LocalDateTime;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.babjo.whatdaybot.model.RisingKeyword;
import com.babjo.whatdaybot.naver.openapi.URLShortenerClient;

public class PeriodicRisingKeywordCrawler {

    private final static Logger logger = LoggerFactory.getLogger(PeriodicRisingKeywordCrawler.class);

    private final Clock clock;
    private final ScheduledExecutorService executorService;
    private final RisingKeywordCrawler risingKeywordCrawler;
    private final URLShortenerClient urlShortenerClient;

    private LocalDateTime latestRefreshTime;
    private List<RisingKeyword> latestRisingKeywords;

    public PeriodicRisingKeywordCrawler(
            ScheduledExecutorService executorService,
            RisingKeywordCrawler risingKeywordCrawler,
            Clock clock,
            URLShortenerClient urlShortenerClient) {
        this.executorService = executorService;
        this.clock = clock;
        this.risingKeywordCrawler = risingKeywordCrawler;
        this.urlShortenerClient = urlShortenerClient;
    }

    public void start() {
        executorService.scheduleAtFixedRate(this::refresh, 0, 60, TimeUnit.SECONDS);
    }

    private void refresh() {
        try {
            CompletableFuture.runAsync(() -> {
                try {
                    LocalDateTime dateTime = LocalDateTime.now(clock);
                    int sec = dateTime.getSecond();
                    sec = sec % 30;
                    dateTime = dateTime.plusSeconds(-sec);

                    List<RisingKeyword> keywords = risingKeywordCrawler.doCrawling(dateTime);

                    latestRefreshTime = dateTime;
                    latestRisingKeywords = keywords.stream().map(
                            keyword -> new RisingKeyword(keyword.getText(), urlShortenerClient
                                    .shorten(keyword.getUrl()))).collect(toImmutableList());

                    logger.info("Success refreshing: {}", latestRisingKeywords);
                } catch (IOException e) {
                    logger.error("Fail to refresh", e);
                }
            }).get(30, TimeUnit.SECONDS);
        } catch (InterruptedException | ExecutionException | TimeoutException e) {
            logger.error("Fail to refresh", e);
        }
    }

    public LocalDateTime getLatestRefreshTime() {
        return latestRefreshTime;
    }

    public List<RisingKeyword> getLatestRisingKeywords() {
        return latestRisingKeywords;
    }
}

File: src/main/kotlin/com/babjo/whatdaybot/crawler/PeriodicRisingKeywordCrawler.kt
package com.babjo.whatdaybot.crawler

import com.babjo.whatdaybot.RisingKeyword
import com.babjo.whatdaybot.naver.openapi.URLShortenerClient
import org.slf4j.LoggerFactory
import java.io.IOException
import java.time.Clock
import java.time.LocalDateTime
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

class PeriodicRisingKeywordCrawler(
    private val clock: Clock,
    private val executorService: ScheduledExecutorService,
    private val risingKeywordCrawler: RisingKeywordCrawler,
    private val urlShortenerClient: URLShortenerClient
) {

    private val logger = LoggerFactory.getLogger(PeriodicRisingKeywordCrawler::class.java)
    lateinit var latestRefreshTime: LocalDateTime
    lateinit var latestRisingKeywords: List<RisingKeyword>

    fun start() {
        executorService.scheduleAtFixedRate(this::refresh, 0, 60, TimeUnit.SECONDS)
    }

    private fun refresh() {
        try {
            CompletableFuture
                .runAsync(this::doRefresh)
                .get(30, TimeUnit.SECONDS)
        } catch (e: Exception) {
            logger.error("Failed to refresh", e)
        }
    }

    private fun doRefresh() {
        try {
            targetDate().also {
                latestRisingKeywords = it
                    .let(risingKeywordCrawler::doCrawling)
                    .map(this::shortenUrl)
                latestRefreshTime = it
            }
        } catch (e: IOException) {
            logger.error("Failed to refresh", e)
        }
    }

    private fun targetDate() =
        LocalDateTime
            .now(clock)
            .let {
                var sec = it.second
                sec %= 30
                it.plusSeconds((-sec).toLong())
            }

    private fun shortenUrl(keyword: RisingKeyword) =
        RisingKeyword(keyword.text, urlShortenerClient.shorten(keyword.url))
}
